File: MERNstack/frontend/src/lib/textbook-service.ts
import { fetchWithAuth } from "./api";

const API_URL = "https://lamp-stack4331.xyz/api/textbooks";

export interface Textbook {
  _id: string;
  title: string;
  author?: string;
  isbn?: string;
  price: number;
  condition: "new" | "like new" | "used" | "very used";
  description?: string;
  images?: string[];
  category?: 
    | "Math"
    | "Science"
    | "Computer Science"
    | "Engineering"
    | "Business"
    | "Literature"
    | "Language"
    | "Uncategorized";
  seller: string; // User ID
  buyer?: string | null; // User ID of buyer, null if unsold
  createdAt: Date;
}

export interface CreateTextbookData {
  title: string;
  author?: string;
  isbn?: string;
  price: number;
  condition: "new" | "like new" | "used" | "very used";
  description?: string;
  images?: string[];
  category?: 
    | "Math"
    | "Science"
    | "Computer Science"
    | "Engineering"
    | "Business"
    | "Literature"
    | "Language"
    | "Uncategorized";
}

export const textbookService = {
  // Get all textbooks by current user
  async getMyTextbooks(userId: string): Promise<Textbook[]> {
    const response = await fetchWithAuth(`${API_URL}/by-user`, {
      method: "POST",
      body: JSON.stringify({ userId }),
    });

    if (!response.ok) {
      throw new Error("Failed to fetch textbooks");
    }

    const data = await response.json();
    return data.textbooks;
  },

  // Add a new textbook
  async addTextbook(textbookData: CreateTextbookData): Promise<Textbook> {
    const response = await fetchWithAuth(`${API_URL}/add`, {
      method: "POST",
      body: JSON.stringify(textbookData),
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.error || "Failed to add textbook");
    }

    return response.json();
  },

  // Update a textbook
  async updateTextbook(
    id: string,
    updates: Partial<CreateTextbookData>
  ): Promise<Textbook> {
    const response = await fetchWithAuth(`${API_URL}/update`, {
      method: "POST",
      body: JSON.stringify({ id, ...updates }),
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.error || "Failed to update textbook");
    }

    const data = await response.json();
    return data.textbook;
  },

  // Delete a textbook
  async deleteTextbook(id: string): Promise<void> {
    const response = await fetchWithAuth(`${API_URL}/delete`, {
      method: "POST",
      body: JSON.stringify({ id }),
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.error || "Failed to delete textbook");
    }
  },

  // Search all textbooks (unsold only)
  async searchTextbooks(query: string): Promise<Textbook[]> {
    const response = await fetch(`${API_URL}/search`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ search: query }),
    });

    if (!response.ok) {
      throw new Error("Search failed");
    }

    const data = await response.json();
    // Filter out sold textbooks
    return data.results.filter((t: Textbook) => !t.buyer);
  },

  // Get all textbooks (unsold only)
  async getAllTextbooks(): Promise<Textbook[]> {
    const response = await fetch(`${API_URL}/all`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
    });

    if (!response.ok) {
      throw new Error("Failed to fetch textbooks");
    }

    const data = await response.json();
    // Filter out sold textbooks
    return data.textbooks.filter((t: Textbook) => !t.buyer);
  },
};

File: MERNstack/mobile/big_project_mobile_app/lib/services/textbook_service.dart
// lib/services/textbook_service.dart

import 'dart:convert';
import 'package:http/http.dart' as http;
import '../api_config.dart';
import 'auth_service.dart';

class Textbook {
  final String id;
  final String title;
  final String author;
  final double price;
  final String condition;
  final List<String> images;
  final String? buyer;
  final String? sellerId;
  final String? sellerName;

  Textbook({
    required this.id,
    required this.title,
    required this.author,
    required this.price,
    required this.condition,
    required this.images,
    this.buyer,
    this.sellerId,
    this.sellerName,
  });

  factory Textbook.fromJson(Map<String, dynamic> json) {
    // images may be array or string
    final rawImages = json['images'];
    List<String> images = [];
    if (rawImages is List) {
      images = rawImages.map((e) => e.toString()).toList();
    } else if (rawImages is String) {
      images = [rawImages];
    }

    // price might be int or double
    double price = 0;
    final rawPrice = json['price'];
    if (rawPrice is num) {
      price = rawPrice.toDouble();
    } else if (rawPrice is String) {
      price = double.tryParse(rawPrice) ?? 0;
    }

    // seller populated from backend: seller: { _id, name, email, phone }
    String? sellerId;
    String? sellerName;
    final seller = json['seller'];
    if (seller is Map<String, dynamic>) {
      sellerId = seller['_id']?.toString();
      sellerName = seller['name']?.toString();
    }

    return Textbook(
      id: (json['_id'] ?? json['id'] ?? '').toString(),
      title: (json['title'] ?? '').toString(),
      author: (json['author'] ?? 'Unknown Author').toString(),
      price: price,
      condition: (json['condition'] ?? 'used').toString(),
      images: images,
      buyer: json['buyer']?.toString(),
      sellerId: sellerId,
      sellerName: sellerName,
    );
  }
}

class TextbookService {
  /// Get all textbooks, then filter out those that already have a buyer.
  Future<List<Textbook>> getAllTextbooks() async {
    final uri = Uri.parse('$apiBaseUrl/api/textbooks/all');

    final res = await http.post(
      uri,
      headers: {'Content-Type': 'application/json'},
    );

    if (res.statusCode == 200) {
      final data = jsonDecode(res.body);
      final list = (data['textbooks'] as List<dynamic>);
      final allBooks =
          list.map((e) => Textbook.fromJson(e as Map<String, dynamic>)).toList();

      // Hide purchased books on the mobile side
      return allBooks
          .where((b) => b.buyer == null || b.buyer!.isEmpty)
          .toList();
    } else {
      throw Exception('Failed to load textbooks (code ${res.statusCode})');
    }
  }

  Future<List<Textbook>> searchTextbooks(String query) async {
    final uri = Uri.parse('$apiBaseUrl/api/textbooks/search');

    final res = await http.post(
      uri,
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({'search': query}),
    );

    if (res.statusCode == 200) {
      final data = jsonDecode(res.body);
      final list = (data['results'] as List<dynamic>);
      final allBooks =
          list.map((e) => Textbook.fromJson(e as Map<String, dynamic>)).toList();

      return allBooks
          .where((b) => b.buyer == null || b.buyer!.isEmpty)
          .toList();
    } else {
      throw Exception('Failed to search textbooks (code ${res.statusCode})');
    }
  }

  /// Books for the currently logged in user (as seller).
  Future<List<Textbook>> getMyTextbooks() async {
    if (AuthService.currentUser == null) {
      throw Exception('Not signed in');
    }
    final user = AuthService.currentUser!;
    final userId =
        (user['_id'] ?? user['id'] ?? '').toString(); // be flexible about key

    final all = await getAllTextbooks();
    return all.where((b) => b.sellerId == userId).toList();
  }

  /// Use existing /api/textbooks/purchase route.
  Future<Map<String, dynamic>> purchaseTextbook(
    String id,
    String shippingAddress,
  ) async {
    if (AuthService.authToken == null) {
      return {'success': false, 'message': 'You must be logged in to buy'};
    }

    final uri = Uri.parse('$apiBaseUrl/api/textbooks/purchase');

    final res = await http.post(
      uri,
      headers: {
        'Content-Type': 'application/json',
        'Authorization': 'Bearer ${AuthService.authToken}',
      },
      body: jsonEncode({'id': id, 'shippingAddress': shippingAddress}),
    );

    if (res.statusCode == 200) {
      return {'success': true};
    } else {
      final data = _safeJson(res.body);
      return {
        'success': false,
        'message': data['error'] ?? 'Failed to purchase',
      };
    }
  }

  /// Create a new textbook listing for the signed-in user.
  Future<Map<String, dynamic>> addTextbook({
    required String title,
    required String author,
    required String isbn,
    required double price,
    required String condition,
    required String description,
    required String imageUrl,
  }) async {
    if (AuthService.authToken == null) {
      return {'success': false, 'message': 'You must be logged in to sell'};
    }

    final uri = Uri.parse('$apiBaseUrl/api/textbooks/add');

    final body = {
      'title': title,
      'author': author,
      'isbn': isbn,
      'price': price,
      'condition': condition,
      'description': description,
      'images': imageUrl.trim().isEmpty ? [] : [imageUrl.trim()],
    };

    final res = await http.post(
      uri,
      headers: {
        'Content-Type': 'application/json',
        'Authorization': 'Bearer ${AuthService.authToken}',
      },
      body: jsonEncode(body),
    );

    final data = _safeJson(res.body);

    if (res.statusCode == 201) {
      return {
        'success': true,
        'textbook': data['textbook'] ?? data,
      };
    } else {
      return {
        'success': false,
        'message': data['error'] ?? 'Failed to create textbook',
      };
    }
  }

  /// Update an existing textbook listing owned by the current user.
  Future<Map<String, dynamic>> updateTextbook({
    required String id,
    required String title,
    required String author,
    required String isbn,
    required double price,
    required String condition,
    required String description,
    required String imageUrl,
  }) async {
    if (AuthService.authToken == null) {
      return {'success': false, 'message': 'You must be logged in to edit'};
    }

    final uri = Uri.parse('$apiBaseUrl/api/textbooks/update');

    final body = {
      'id': id,
      'title': title,
      'author': author,
      'isbn': isbn,
      'price': price,
      'condition': condition,
      'description': description,
      'images': imageUrl.trim().isEmpty ? [] : [imageUrl.trim()],
    };

    final res = await http.post(
      uri,
      headers: {
        'Content-Type': 'application/json',
        'Authorization': 'Bearer ${AuthService.authToken}',
      },
      body: jsonEncode(body),
    );

    final data = _safeJson(res.body);

    if (res.statusCode == 200) {
      return {
        'success': true,
        'textbook': data['textbook'] ?? data,
      };
    } else {
      return {
        'success': false,
        'message': data['error'] ?? 'Failed to update textbook',
      };
    }
  }

  /// Delete an existing textbook listing owned by the current user.
  Future<Map<String, dynamic>> deleteTextbook(String id) async {
    if (AuthService.authToken == null) {
      return {'success': false, 'message': 'You must be logged in to delete'};
    }

    final uri = Uri.parse('$apiBaseUrl/api/textbooks/delete');

    final res = await http.post(
      uri,
      headers: {
        'Content-Type': 'application/json',
        'Authorization': 'Bearer ${AuthService.authToken}',
      },
      body: jsonEncode({'id': id}),
    );

    final data = _safeJson(res.body);

    if (res.statusCode == 200) {
      return {'success': true};
    } else {
      return {
        'success': false,
        'message': data['error'] ?? 'Failed to delete textbook',
      };
    }
  }

  Map<String, dynamic> _safeJson(String body) {
    try {
      final v = jsonDecode(body);
      return (v is Map<String, dynamic>) ? v : <String, dynamic>{};
    } catch (_) {
      return <String, dynamic>{};
    }
  }
}
